src/request.rs

```diff
@@ -377,23 +376,52 @@ impl Request {
     ///     .query("foo", "42")
     ///     .query("foo", "43");
     ///
-    /// assert_eq!(req.query_params(), vec![
-    ///     ("foo".to_string(), "42".to_string()),
-    ///     ("foo".to_string(), "43".to_string())
+    /// assert_eq!(req.query_pairs(), vec![
+    ///     ("foo", "42"),
+    ///     ("foo", "43")
     /// ]);
     /// # Ok(())
     /// # }
     /// ```
-    pub fn query_params(&self) -> Vec<(String, String)> {
-        let mut ret = vec![];
-
-        if let Ok(url) = &self.parsed_url.0 {
-            for (k, v) in url.query_pairs() {
-                ret.push((k.into(), v.into()));
-            }
-        }
+    pub fn query_pairs(&self) -> Vec<(&str, &str)> {
+        self.query_pairs
+            .iter()
+            .map(|(k, v)| (k.as_str(), v.as_str()))
+            .collect()
+    }
 
-        ret
+    /// Get the url str that will be used for this request.
+    ///
+    /// The url might differ from that originally provided when constructing the
+    /// request if additional query parameters have been added using [`Request::query()`].
+    ///
+    /// In case the original url provided to build the request is not possible to
+    /// parse to a Url, this function returns the original, and it will error once the
+    /// Request object is used.
```

Comment:
```suggestion
    /// In case the original url provided to build the request is not possible to
    /// parse to a Url, this function returns the original, and the Request will
    /// produce an error when `call` or `send_*` is called.
```